### PR TITLE
take subtracted type into account in `MixedType->isArray()`

### PR DESCRIPTION
**Pull request by staabm**

### CODE REVIEW

File: src/Type/MixedType.php [hunk 1]
```diff
@@ -408,6 +408,22 @@ public function traverse(callable $cb): Type
 
 	public function isArray(): TrinaryLogic
 	{
+		if ($this->subtractedType !== null) {
+			$arrayType = new ArrayType(new MixedType(), new MixedType());
+
+			if ($arrayType->isSuperTypeOf($this->subtractedType)->yes()) {
```
Comment (ondrejmirtes) [inline src/Type/MixedType.php]: You need to switch this around.


No other code is needed for this to work, no instanceof UnionType etc.

File: tests/PHPStan/Type/MixedTypeTest.php [hunk 2]
```diff
@@ -164,4 +166,73 @@ public function testIsSuperTypeOf(MixedType $type, Type $otherType, TrinaryLogic
 		);
 	}
 
+	public function dataSubstractedIsArray(): array
+	{
+		return [
+			[
+				new MixedType(),
+				new ArrayType(new IntegerType(), new StringType()),
+				TrinaryLogic::createNo(),
```
Comment (ondrejmirtes) [inline tests/PHPStan/Type/MixedTypeTest.php]: These tests are wrong. It can still be another array.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: tests/PHPStan/Type/MixedTypeTest.php (hunk 2 above)
Comment (staabm) [inline tests/PHPStan/Type/MixedTypeTest.php]: subtracting only `array<int, string>` could still leave e.g. a `array<string, mixed>` within the `mixed`.
therefore I am wondering, whether the only case where `$mixed->isArray()` can say NO is, when a general `array` is substracted?

Comment (ondrejmirtes) [reply tests/PHPStan/Type/MixedTypeTest.php]: Exactly. Switching the callee and the argument works even when more than just an array are subtracted, like array|null for example.

Comment (staabm) [reply tests/PHPStan/Type/MixedTypeTest.php]: I just see https://github.com/phpstan/phpstan-src/pull/1656#discussion_r957025938, so I think we are on the same page.. will adjust the PR.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
While you're at it, please do all the other methods as well.

**Comment by staabm** [comment]
implemented in all methods I found useful.

here are a few notes:
- should `isIterableAtLeastOnce` return `NO` when `isIterable` returns `NO`?
- was not sure what todo with `isOffsetAccessible` and friends ..?
- was wondering whether `isCallable` should also use a `$this->subtractedType->isSuperTypeOf(new CallableType(...))` construction instead?
see https://github.com/phpstan/phpstan-src/blob/9ff3322395b703850111478df3369bb2bbf07438/src/Type/MixedType.php#L130-L140

**Comment by ondrejmirtes** [comment]
> should `isIterableAtLeastOnce` return `NO` when `isIterable` returns `NO`?

Exactly, it should just call `return $this->isIterable()`. I don't think we can get more precise than that.

> - was not sure what todo with `isOffsetAccessible` and friends ..?

Can we get a union of all offset-accessible types? Is it `string|array|ArrayAccess`? Yeah, probably.

> - was wondering whether `isCallable` should also use a `$this->subtractedType->isSuperTypeOf(new CallableType(...))` construction instead?

Yes, definitely, the current code is wrong.

**Comment by staabm** [comment]
implemented and rebased

**Comment by ondrejmirtes** [comment]
Thank you.

**Pull request merged**